core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -566,42 +556,61 @@ void finishOnRemoteSuccess(Response response) {
-            final String writeConsistencyFailure = checkWriteConsistency(primary);
+        void retryBecauseUnavailable(ShardId shardId, String message) {
+            retry(new UnavailableShardsException(shardId, message + " Timeout: [" + request.timeout() + "], request: " + request.toString()));
+        }
+    }
+
+    /**
+     * Responsible for performing primary operation locally and delegating to replication action once successful
+     * <p>
+     * Note that as soon as we move to replication action, state responsibility is transferred to {@link ReplicationPhase}.
+     */
+    final class PrimaryPhase extends AbstractRunnable {
+        private final Request request;
+        private final TransportChannel channel;
+        private final ClusterState clusterState;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
+
+        PrimaryPhase(Request request, TransportChannel channel) {
+            this.clusterState = clusterService.state();
+            this.request = request;
+            this.channel = channel;
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            finishAsFailed(e);
+        }
+
+        @Override
+        protected void doRun() throws Exception {
+            // request shardID was set in ReroutePhase
+            assert request.shardId() != null : "request shardID must be set prior to primary phase";
+            final ShardId shardId = request.shardId();
+            final String writeConsistencyFailure = checkWriteConsistency(shardId);
             if (writeConsistencyFailure != null) {
-                retryBecauseUnavailable(primary.shardId(), writeConsistencyFailure);
+                finishBecauseUnavailable(shardId, writeConsistencyFailure);
                 return;
             }
             final ReplicationPhase replicationPhase;
             try {
-                indexShardReference = getIndexShardOperationsCounter(primary.shardId());
-                PrimaryOperationRequest por = new PrimaryOperationRequest(primary.id(), internalRequest.concreteIndex(), internalRequest.request());
-                Tuple<Response, ReplicaRequest> primaryResponse = shardOperationOnPrimary(observer.observedState(), por);
-                logger.trace("operation completed on primary [{}]", primary);
-                replicationPhase = new ReplicationPhase(shardsIt, primaryResponse.v2(), primaryResponse.v1(), observer, primary, internalRequest, listener, indexShardReference, shardFailedTimeout);
+                indexShardReference = getIndexShardOperationsCounter(shardId);
+                Tuple<Response, ReplicaRequest> primaryResponse = shardOperationOnPrimary(clusterState, request);
+                logger.trace("operation completed on primary [{}]", shardId);
+                // we cache meta data to resolve settings even if the index gets deleted after primary operation
```

Comment:
This was to prevent a NPE on index metadata used to decide whether we wanted to execute actions on replica (for shadow replicas) and grab the indexUUID at the replication phase. This is possible when the index has been deleted after the primary operation has completed and before the replication phase has begun. 

I have cleaned this up and now the case is handled in ReplicationPhase ctor. Now if the index gets deleted before replication, the replication phase is skipped.